### DISCUSSION TITLE
HTML renderer not rendering custom components in NextJS

### DISCUSSION THREAD

**Discussion by ian-whitestone** [markdoc/markdoc#370] [Q&A] [answered]
# Background

I'm using Markdoc to manage a product changelog on a React/NextJS website.

Here's the current setup:

1. Example changelog post which uses a custom tag:

```md
# src/pages/changelog/test.md
---
title: Example feature
date: 2023-04-11T12:00:00.000Z
---

Words more words..

{% custom-thing /%}

```

2. The component used in the custom tag:

```js
// src/components/markdoc/CustomThing.js
export function CustomThing() {
  return (
    <>
      <p>
        Bunch of text
      </p>
    </>
  )
}

```

3. All the posts get collected in a server side function

```js
// src/utils/getAllPosts.js

import { CustomThing } from '@/components/markdoc/CustomThing.js'

export function getAllPosts() {
  const files = glob.sync(`src/pages/changelog/test.md`)
  const posts = []

  const config = {
    tags: {
      'custom-thing': {
        selfClosing: true,
        render: CustomThing,
      },
    },
  }

  files.forEach((file) => {
    const rawText = fs.readFileSync(file, 'utf-8')
    const ast = Markdoc.parse(rawText)
    const content = Markdoc.transform(ast, config)
    const html = Markdoc.renderers.html(content)

    const post = {
      html: html
    }
    
    posts.push(post)
  })

  return posts
}
```

4. The posts get rendered on the page

```js
// src/pages/changelog/index.js

export default function Changelog({ posts }) {
  return (
    <>
      <div>
        {posts.map((post) => (
          <article>
              <div dangerouslySetInnerHTML={{ __html: post.html }} />
          </article>
        ))}
      </div>
    </>
  )
}

export async function getStaticProps() {
  const posts = getAllPosts()
  return {
    props: {
      posts,
    },
  }
}
```

# The Problem

When rendering the html on the server side function, the `custom-thing` tag is not converted to the underlying html from the component. Is there anything you can spot that I'm doing incorrectly? Happy to provide more details on the setup, versions, etc. if required.

```js
  const config = {
    tags: {
      'custom-thing': {
        selfClosing: true,
        render: CustomThing,
      },
    },
  }

    const ast = Markdoc.parse(rawText)
    const content = Markdoc.transform(ast, config)
    const html = Markdoc.renderers.html(content)
```

![CleanShot 2023-05-08 at 23 45 42@2x](https://user-images.githubusercontent.com/11096727/236988916-f7a9b647-6cc6-4fed-8af5-ef9c01b8cff8.png)

Note, the tag and changelog post render as expected when visiting the underlying page (i.e. `/changelog/test` where NextJS plugin renders `test.md` as expected with the custom component).

**Comment by mfix-stripe** [comment]
@ian-whitestone looks like you are using React…if so, why not use the [Markdoc React renderer](https://markdoc.dev/docs/examples/react)?

You can call `Markdoc.transform` on the server, and then call `Markdoc.renderers.react(content, React, { components })` on the client  (replacing  `<div dangerouslySetInnerHTML={{ __html: post.html }} />`)

**Reply by ian-whitestone** [reply]
@mfix-stripe thanks for the quick response!

I actually went down that path before trying the html renderer. The challenge i faced was that for NextJS, we have our custom nodes / tags defined like so:

```js
// markdoc/tags.js
const tags = {
  callout: {
    attributes: {
      title: { type: String },
      type: {
        type: String,
        default: 'note',
        matches: ['note', 'warning'],
        errorLevel: 'critical',
      },
    },
    render: Callout,
  },
  ... // etc...
}

export default tags
```

```js
// markdoc/nodes.js
const nodes = {
  fence: {
    render: Fence,
    attributes: {
      language: {
        type: String,
      },
      highlights: {
        type: String,
      },
    },
  },
  ... // etc...
}

export default nodes
```

But in order to use the React renderer, all the components must be strings, i.e. instead of `fence: { render: Fence, ...}` and `callout: {render: Callout, ...}`, I had to create/maintain separate variables like `const serializableNodes = { fence: { render: 'Fence', ...}, ...}` and `const serializableTags = { callout: { render: 'Callout', ...}, ...}`, and then also create/maintain a component mapping like `const componentsMap =  { 'Callout': Callout, 'Fence': Fence, ... }`, which then all got fed into:

```js
Markdoc.renderers.react(content, React, {
    components: componentMap,
  })
```

After a lot of fiddling this ended up working fine in local development, but after running `next build` it stopped working and failed with this error: https://legacy.reactjs.org/docs/error-decoder.html/?args%5B%5D=undefined&args%5B%5D=&invariant=130

<img width="1481" alt="CleanShot 2023-05-09 at 14 22 15@2x" src="https://github.com/markdoc/markdoc/assets/11096727/3d4c0b15-68a9-4ef8-bb55-2005faafbbdf">

(you can actually visit the [currently broken page](https://select.dev/changelog) if interested to see the console i screenshotted above).

This was quite tricky to debug:

> Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined.

So that's why i went down the approach of trying to use the html renderer. Does the html renderer not work with NextJS? Or alternatively, do you have any idea why the Markdoc React renderer approach failed when building the static NextJS page?

Any ideas/advice would be greatly appreciated!

**Reply by ian-whitestone** [reply]
@mfix-stripe just wanted to follow up on this in case it got lost since this is marked as closed

**Reply by mfix-stripe** [reply, marked as answer]
Hey @ian-whitestone, you can follow the [pattern here](https://github.com/markdoc/docs/blob/main/components/Sandbox.js#L45)—calling `getSchema` from `@markdoc/next.js/runtime`
